<a href="https://colab.research.google.com/github/Dieljr/README/blob/main/GDS_Grid.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Conteúdo para aulas online
## Terceiro Notebook: GridSearchCV para SGDR

### O Objetivo desse notebook é o de apresentar recursos intermediários a avançados sobre o Algoritmo GridSearchCV ou Busca em Gride com Validação Cruzada do SKLearn, da classe dos seletores de modelos (Model Selection).

1. Source: 

a) <font color = '#CD5C5C'>Site do sklearn:</font> https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html?highlight=grid#sklearn.model_selection.GridSearchCV

b) <font colo = '#CD5C5C'>Livro Mãos à Obra: Aprendizado de Máquina com Scikit-Learn e TensorFlow</font>: https://www.amazon.com.br/M%C3%A3os-obra-aprendizado-Scikit-Learn-inteligentes/dp/8550815489/ref=sr_1_1?crid=2JG8SYAU8K3UQ&keywords=tensorflow&qid=1646567493&s=books&sprefix=tensort+flow%2Cstripbooks%2C229&sr=1-1&ufe=app_do%3Aamzn1.fos.6d798eae-cadf-45de-946a-f477d47705b9

c) <font color = '#CD5C5C'>Livro Análise de dados com Python e Pandas</font>:https://www.amazon.com.br/An%C3%A1lise-Dados-com-Python-Pandas/dp/8575226991/ref=sr_1_1?__mk_pt_BR=%C3%85M%C3%85%C5%BD%C3%95%C3%91&keywords=analise+de+dados+com+python&qid=1646671469&sr=8-1&ufe=app_do%3Aamzn1.fos.6d798eae-cadf-45de-946a-f477d47705b9


3. Parametros ou hiperparamentros:
Para você que está começando, pode parecer estranho e confuso, mas é assim que a comunidade em Ciência de Dados se refere ajustadores dos algoritmos de machine learning. Então, vá se acostumando!
Irei apresentar somente os mais importantes parametros que o ajudará no seu dia-a-dia como cientista de dados.

class sklearn.model_selection.GridSearchCV(estimator, param_grid, *, scoring=None, n_jobs=None, refit=True, cv=None, verbose=0, pre_dispatch='2*n_jobs', error_score=nan, return_train_score=False)

a) cv: por padrão é 'None'. Trata-se da Cross Validation.

b) verbose: por padrão é 0. Apresenta os resultados do processamento.

c) param_grid: é o local aonde guardamos a relação de parâmetros do outro algoritmo. É o mais IMPORTANTE item para esse estudo.

d) scoring: modo usado para avaliar o cross validation. Exemplos: explained_variance; neg_mean_absolute_error; neg_mean_squared_error;
r2; neg_mean_poisson_deviance etc.Para mais informações veja site: https://scikit-learn.org/stable/modules/model_evaluation.html#scoring-parameter


Atributos: ou o que se pode extrair do algoritmo depois que ele for usado. Melhor do que explicá-los aqui (teoria), é apresentá-los na prática!

Cometários:
O GridSearchCV é um algoritmo que, em geral, ajuda outros algoritmos de REGRESSÃO e CLASSIFICAÇÃO a encontrar os seus melhores parâmetros.
O Grid (para os íntimos) faz uma busca exaustiva de combinações até apresentar os melhores valores.
Por si só, ele é capaz de realizar o FIT, o PREDICT e o SCORE sem auxílio de outros algoritmos.

## PRÁTICA:

In [ ]:
# Carregar pacotes

import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import matplotlib as mpl
import seaborn as sb

from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import SGDRegressor
from sklearn.model_selection import train_test_split, GridSearchCV

import sys

In [ ]:
print(sys.version)

3.7.6 (default, Jan  8 2020, 20:23:39) [MSC v.1916 64 bit (AMD64)]


In [ ]:
# Carregar dataset
df = pd.read_csv('mpg.csv', sep = ';', names=['mpg', 'cylinders', 'displacement', 'horsepower', 'weight', 'accelaration',
                                              'model_year', 'origin', 'car_name'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 406 entries, 0 to 405
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     406 non-null    float64
 2   displacement  406 non-null    float64
 3   horsepower    400 non-null    float64
 4   weight        406 non-null    float64
 5   accelaration  406 non-null    float64
 6   model_year    406 non-null    float64
 7   origin        1 non-null      object 
 8   car_name      0 non-null      float64
dtypes: float64(8), object(1)
memory usage: 28.7+ KB


### Pequenos ajustes

In [ ]:
# Não precisaremos das duas últimas colunas
del df['origin']
del df['car_name']

In [ ]:
df.head()

mpg  cylinders  displacement  horsepower  weight  accelaration  model_year
0  18.0        8.0         307.0       130.0  3504.0          12.0        70.0
1  15.0        8.0         350.0       165.0  3693.0          11.5        70.0
2  18.0        8.0         318.0       150.0  3436.0          11.0        70.0
3  16.0        8.0         304.0       150.0  3433.0          12.0        70.0
4  17.0        8.0         302.0       140.0  3449.0          10.5        70.0

In [ ]:
# Podemos excluir os valores missing sem perda significativa.
df = df.dropna()

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 392 entries, 0 to 405
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           392 non-null    float64
 1   cylinders     392 non-null    float64
 2   displacement  392 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        392 non-null    float64
 5   accelaration  392 non-null    float64
 6   model_year    392 non-null    float64
dtypes: float64(7)
memory usage: 24.5 KB


In [ ]:
# Separar as variáveis para o algoritmo
X = df.drop(['mpg'], axis = 1)
y = df['mpg']

In [ ]:
# Normalizar o conjunto das features, pois o algoritmo exige para dados não normalizados
scale = StandardScaler()
X_scale = scale.fit_transform(X)

In [ ]:
# Dividir em dados treino e teste
X_treino, X_teste, y_treino, y_teste = train_test_split(X_scale, y, test_size = 0.3, random_state = 100)

### Avaliação com diferentes configurações dos hiperparametros com GridSearchCV

### Machine Learning com SGDRegressor e GridSearchCV

In [ ]:
# Criar objeto SGDRegressor, apenas alterando verbose para acompanharmos as info geradas pelo algoritmo.
sgdr = SGDRegressor(random_state=590, verbose=True)

# Criar parâmetros para o SGDR com o recurso do Grid (param_grid) em forma de dicionário
param_grid = ({'penalty': ['L1', 'L2'], 
               'alpha': [0.0001, 0.001, 0.005, 0.01],
               'eta0': [0.01, 0.05, 0.1]})

# Criar objeto GridSearchCV
grid = GridSearchCV(sgdr,
                    param_grid,
                    cv = 10,
                    scoring = 'r2')

In [ ]:
# treinar o modelo
import datetime

inicio = datetime.timedelta()

grid.fit(X_scale, y)

fim = datetime.timedelta()

print(fim - inicio)


-- Epoch 1
Norm: 4.49, NNZs: 6, Bias: 15.188947, T: 352, Avg. loss: 103.084620
Total training time: 0.00 seconds.
-- Epoch 2
Norm: 4.88, NNZs: 6, Bias: 19.203077, T: 704, Avg. loss: 22.870610
Total training time: 0.00 seconds.
-- Epoch 3
Norm: 4.82, NNZs: 6, Bias: 20.980973, T: 1056, Avg. loss: 11.055889
Total training time: 0.00 seconds.
-- Epoch 4
Norm: 4.59, NNZs: 6, Bias: 21.883414, T: 1408, Avg. loss: 8.004745
Total training time: 0.00 seconds.
-- Epoch 5
Norm: 4.62, NNZs: 6, Bias: 22.377203, T: 1760, Avg. loss: 7.108758
Total training time: 0.00 seconds.
-- Epoch 6
Norm: 4.58, NNZs: 6, Bias: 22.643085, T: 2112, Avg. loss: 6.763259
Total training time: 0.00 seconds.
-- Epoch 7
Norm: 4.56, NNZs: 6, Bias: 22.816215, T: 2464, Avg. loss: 6.610422
Total training time: 0.00 seconds.
-- Epoch 8
Norm: 4.60, NNZs: 6, Bias: 22.923026, T: 2816, Avg. loss: 6.525167
Total training time: 0.00 seconds.
-- Epoch 9
Norm: 4.63, NNZs: 6, Bias: 22.969577, T: 3168, Avg. loss: 6.473397
Total training t

Norm: 5.38, NNZs: 6, Bias: 23.168036, T: 22880, Avg. loss: 6.146265
Total training time: 0.05 seconds.
-- Epoch 66
Norm: 5.39, NNZs: 6, Bias: 23.171426, T: 23232, Avg. loss: 6.145513
Total training time: 0.05 seconds.
-- Epoch 67
Norm: 5.40, NNZs: 6, Bias: 23.172009, T: 23584, Avg. loss: 6.141476
Total training time: 0.05 seconds.
-- Epoch 68
Norm: 5.38, NNZs: 6, Bias: 23.169127, T: 23936, Avg. loss: 6.131215
Total training time: 0.05 seconds.
-- Epoch 69
Norm: 5.41, NNZs: 6, Bias: 23.169819, T: 24288, Avg. loss: 6.135910
Total training time: 0.05 seconds.
-- Epoch 70
Norm: 5.40, NNZs: 6, Bias: 23.166417, T: 24640, Avg. loss: 6.136134
Total training time: 0.05 seconds.
-- Epoch 71
Norm: 5.43, NNZs: 6, Bias: 23.174340, T: 24992, Avg. loss: 6.138360
Total training time: 0.05 seconds.
-- Epoch 72
Norm: 5.45, NNZs: 6, Bias: 23.179571, T: 25344, Avg. loss: 6.135511
Total training time: 0.06 seconds.
-- Epoch 73
Norm: 5.45, NNZs: 6, Bias: 23.181574, T: 25696, Avg. loss: 6.136827
Total traini

Norm: 5.87, NNZs: 6, Bias: 23.430549, T: 3177, Avg. loss: 6.073487
Total training time: 0.00 seconds.
-- Epoch 10
Norm: 6.20, NNZs: 6, Bias: 23.426681, T: 3530, Avg. loss: 5.983776
Total training time: 0.00 seconds.
-- Epoch 11
Norm: 6.14, NNZs: 6, Bias: 23.469085, T: 3883, Avg. loss: 6.049602
Total training time: 0.00 seconds.
-- Epoch 12
Norm: 6.12, NNZs: 6, Bias: 23.435843, T: 4236, Avg. loss: 6.059656
Total training time: 0.00 seconds.
-- Epoch 13
Norm: 5.98, NNZs: 6, Bias: 23.429913, T: 4589, Avg. loss: 5.984664
Total training time: 0.00 seconds.
-- Epoch 14
Norm: 6.10, NNZs: 6, Bias: 23.340885, T: 4942, Avg. loss: 6.041979
Total training time: 0.00 seconds.
-- Epoch 15
Norm: 6.02, NNZs: 6, Bias: 23.384751, T: 5295, Avg. loss: 6.035483
Total training time: 0.00 seconds.
Convergence after 15 epochs took 0.00 seconds
-- Epoch 1
Norm: 4.17, NNZs: 6, Bias: 22.854628, T: 353, Avg. loss: 23.809917
Total training time: 0.00 seconds.
-- Epoch 2
Norm: 4.51, NNZs: 6, Bias: 23.008650, T: 706

Norm: 5.42, NNZs: 6, Bias: 23.171933, T: 23232, Avg. loss: 6.144111
Total training time: 0.02 seconds.
-- Epoch 67
Norm: 5.44, NNZs: 5, Bias: 23.172453, T: 23584, Avg. loss: 6.139933
Total training time: 0.02 seconds.
-- Epoch 68
Norm: 5.43, NNZs: 6, Bias: 23.169885, T: 23936, Avg. loss: 6.131514
Total training time: 0.02 seconds.
-- Epoch 69
Norm: 5.45, NNZs: 6, Bias: 23.170194, T: 24288, Avg. loss: 6.133978
Total training time: 0.02 seconds.
-- Epoch 70
Norm: 5.45, NNZs: 6, Bias: 23.166562, T: 24640, Avg. loss: 6.133097
Total training time: 0.02 seconds.
-- Epoch 71
Norm: 5.47, NNZs: 6, Bias: 23.174343, T: 24992, Avg. loss: 6.135900
Total training time: 0.02 seconds.
-- Epoch 72
Norm: 5.49, NNZs: 5, Bias: 23.179581, T: 25344, Avg. loss: 6.134757
Total training time: 0.02 seconds.
-- Epoch 73
Norm: 5.50, NNZs: 6, Bias: 23.181652, T: 25696, Avg. loss: 6.134661
Total training time: 0.02 seconds.
Convergence after 73 epochs took 0.02 seconds
-- Epoch 1
Norm: 4.20, NNZs: 6, Bias: 15.52963

Norm: 4.04, NNZs: 6, Bias: 22.486954, T: 1412, Avg. loss: 7.987601
Total training time: 0.00 seconds.
-- Epoch 5
Norm: 4.21, NNZs: 6, Bias: 22.956680, T: 1765, Avg. loss: 7.087079
Total training time: 0.00 seconds.
-- Epoch 6
Norm: 4.18, NNZs: 6, Bias: 23.241643, T: 2118, Avg. loss: 6.797269
Total training time: 0.00 seconds.
-- Epoch 7
Norm: 4.29, NNZs: 6, Bias: 23.387226, T: 2471, Avg. loss: 6.668247
Total training time: 0.00 seconds.
-- Epoch 8
Norm: 4.28, NNZs: 6, Bias: 23.458001, T: 2824, Avg. loss: 6.606569
Total training time: 0.00 seconds.
-- Epoch 9
Norm: 4.34, NNZs: 6, Bias: 23.512861, T: 3177, Avg. loss: 6.577376
Total training time: 0.00 seconds.
-- Epoch 10
Norm: 4.56, NNZs: 6, Bias: 23.550087, T: 3530, Avg. loss: 6.474935
Total training time: 0.00 seconds.
-- Epoch 11
Norm: 4.61, NNZs: 6, Bias: 23.564122, T: 3883, Avg. loss: 6.499792
Total training time: 0.00 seconds.
-- Epoch 12
Norm: 4.53, NNZs: 6, Bias: 23.571137, T: 4236, Avg. loss: 6.527964
Total training time: 0.00 

Total training time: 0.00 seconds.
-- Epoch 11
Norm: 5.91, NNZs: 6, Bias: 23.539709, T: 3883, Avg. loss: 6.311355
Total training time: 0.00 seconds.
-- Epoch 12
Norm: 5.90, NNZs: 6, Bias: 23.526467, T: 4236, Avg. loss: 6.399368
Total training time: 0.00 seconds.
-- Epoch 13
Norm: 5.81, NNZs: 6, Bias: 23.587587, T: 4589, Avg. loss: 6.329784
Total training time: 0.00 seconds.
-- Epoch 14
Norm: 5.92, NNZs: 6, Bias: 23.495228, T: 4942, Avg. loss: 6.334885
Total training time: 0.00 seconds.
-- Epoch 15
Norm: 5.79, NNZs: 6, Bias: 23.412547, T: 5295, Avg. loss: 6.337001
Total training time: 0.00 seconds.
Convergence after 15 epochs took 0.00 seconds
-- Epoch 1
Norm: 4.67, NNZs: 6, Bias: 23.810539, T: 353, Avg. loss: 25.214917
Total training time: 0.00 seconds.
-- Epoch 2
Norm: 4.57, NNZs: 6, Bias: 23.626579, T: 706, Avg. loss: 6.018746
Total training time: 0.00 seconds.
-- Epoch 3
Norm: 4.91, NNZs: 6, Bias: 23.470862, T: 1059, Avg. loss: 5.924074
Total training time: 0.00 seconds.
-- Epoch 4


-- Epoch 50
Norm: 5.64, NNZs: 6, Bias: 23.580304, T: 17650, Avg. loss: 6.219044
Total training time: 0.03 seconds.
-- Epoch 51
Norm: 5.70, NNZs: 5, Bias: 23.591241, T: 18003, Avg. loss: 6.231660
Total training time: 0.03 seconds.
-- Epoch 52
Norm: 5.69, NNZs: 5, Bias: 23.575120, T: 18356, Avg. loss: 6.215850
Total training time: 0.03 seconds.
-- Epoch 53
Norm: 5.73, NNZs: 5, Bias: 23.579368, T: 18709, Avg. loss: 6.222787
Total training time: 0.03 seconds.
-- Epoch 54
Norm: 5.77, NNZs: 5, Bias: 23.578560, T: 19062, Avg. loss: 6.214616
Total training time: 0.03 seconds.
-- Epoch 55
Norm: 5.79, NNZs: 5, Bias: 23.578628, T: 19415, Avg. loss: 6.217081
Total training time: 0.03 seconds.
-- Epoch 56
Norm: 5.81, NNZs: 5, Bias: 23.576387, T: 19768, Avg. loss: 6.217802
Total training time: 0.03 seconds.
-- Epoch 57
Norm: 5.81, NNZs: 5, Bias: 23.573371, T: 20121, Avg. loss: 6.216124
Total training time: 0.03 seconds.
-- Epoch 58
Norm: 5.86, NNZs: 5, Bias: 23.581092, T: 20474, Avg. loss: 6.200552


Total training time: 0.02 seconds.
-- Epoch 43
Norm: 5.12, NNZs: 6, Bias: 23.625315, T: 15179, Avg. loss: 5.612590
Total training time: 0.02 seconds.
-- Epoch 44
Norm: 5.16, NNZs: 6, Bias: 23.628180, T: 15532, Avg. loss: 5.606227
Total training time: 0.02 seconds.
-- Epoch 45
Norm: 5.19, NNZs: 6, Bias: 23.633339, T: 15885, Avg. loss: 5.605621
Total training time: 0.02 seconds.
-- Epoch 46
Norm: 5.17, NNZs: 6, Bias: 23.629355, T: 16238, Avg. loss: 5.607214
Total training time: 0.02 seconds.
-- Epoch 47
Norm: 5.18, NNZs: 6, Bias: 23.622388, T: 16591, Avg. loss: 5.602337
Total training time: 0.02 seconds.
-- Epoch 48
Norm: 5.20, NNZs: 6, Bias: 23.624672, T: 16944, Avg. loss: 5.602922
Total training time: 0.03 seconds.
-- Epoch 49
Norm: 5.20, NNZs: 6, Bias: 23.627427, T: 17297, Avg. loss: 5.599217
Total training time: 0.03 seconds.
-- Epoch 50
Norm: 5.18, NNZs: 6, Bias: 23.633344, T: 17650, Avg. loss: 5.584528
Total training time: 0.03 seconds.
-- Epoch 51
Norm: 5.24, NNZs: 6, Bias: 23.641

-- Epoch 1
Norm: 6.29, NNZs: 6, Bias: 24.432891, T: 353, Avg. loss: 23.708752
Total training time: 0.00 seconds.
-- Epoch 2
Norm: 6.01, NNZs: 6, Bias: 23.312005, T: 706, Avg. loss: 6.770273
Total training time: 0.00 seconds.
-- Epoch 3
Norm: 6.37, NNZs: 6, Bias: 23.437515, T: 1059, Avg. loss: 6.709585
Total training time: 0.00 seconds.
-- Epoch 4
Norm: 6.64, NNZs: 6, Bias: 23.337286, T: 1412, Avg. loss: 6.545636
Total training time: 0.00 seconds.
-- Epoch 5
Norm: 6.65, NNZs: 6, Bias: 23.483866, T: 1765, Avg. loss: 6.534331
Total training time: 0.00 seconds.
-- Epoch 6
Norm: 6.69, NNZs: 5, Bias: 23.801887, T: 2118, Avg. loss: 6.559632
Total training time: 0.00 seconds.
-- Epoch 7
Norm: 6.92, NNZs: 6, Bias: 23.581122, T: 2471, Avg. loss: 6.489415
Total training time: 0.00 seconds.
-- Epoch 8
Norm: 6.81, NNZs: 5, Bias: 23.284966, T: 2824, Avg. loss: 6.574450
Total training time: 0.00 seconds.
-- Epoch 9
Norm: 7.06, NNZs: 6, Bias: 23.642465, T: 3177, Avg. loss: 6.531353
Total training time

-- Epoch 1
Norm: 2.45, NNZs: 6, Bias: 14.975992, T: 353, Avg. loss: 102.411395
Total training time: 0.00 seconds.
-- Epoch 2
Norm: 2.67, NNZs: 6, Bias: 19.107189, T: 706, Avg. loss: 23.350045
Total training time: 0.00 seconds.
-- Epoch 3
Norm: 3.09, NNZs: 5, Bias: 20.935220, T: 1059, Avg. loss: 10.381935
Total training time: 0.00 seconds.
-- Epoch 4
Norm: 3.46, NNZs: 6, Bias: 21.852815, T: 1412, Avg. loss: 6.940664
Total training time: 0.00 seconds.
-- Epoch 5
Norm: 3.74, NNZs: 6, Bias: 22.373295, T: 1765, Avg. loss: 5.787959
Total training time: 0.00 seconds.
-- Epoch 6
Norm: 3.80, NNZs: 6, Bias: 22.683387, T: 2118, Avg. loss: 5.380452
Total training time: 0.00 seconds.
-- Epoch 7
Norm: 3.96, NNZs: 6, Bias: 22.847539, T: 2471, Avg. loss: 5.196153
Total training time: 0.00 seconds.
-- Epoch 8
Norm: 4.06, NNZs: 6, Bias: 22.952619, T: 2824, Avg. loss: 5.099841
Total training time: 0.00 seconds.
-- Epoch 9
Norm: 4.12, NNZs: 6, Bias: 23.013975, T: 3177, Avg. loss: 5.048513
Total training t

Norm: 5.67, NNZs: 6, Bias: 23.439942, T: 1408, Avg. loss: 6.132848
Total training time: 0.00 seconds.
-- Epoch 5
Norm: 6.07, NNZs: 5, Bias: 23.333320, T: 1760, Avg. loss: 6.284795
Total training time: 0.00 seconds.
-- Epoch 6
Norm: 6.21, NNZs: 5, Bias: 23.228625, T: 2112, Avg. loss: 6.235756
Total training time: 0.00 seconds.
-- Epoch 7
Norm: 6.21, NNZs: 5, Bias: 23.404983, T: 2464, Avg. loss: 6.184305
Total training time: 0.00 seconds.
-- Epoch 8
Norm: 6.60, NNZs: 6, Bias: 23.222906, T: 2816, Avg. loss: 6.103434
Total training time: 0.00 seconds.
-- Epoch 9
Norm: 6.70, NNZs: 6, Bias: 23.057582, T: 3168, Avg. loss: 6.124486
Total training time: 0.00 seconds.
-- Epoch 10
Norm: 6.55, NNZs: 4, Bias: 23.267366, T: 3520, Avg. loss: 6.109350
Total training time: 0.00 seconds.
-- Epoch 11
Norm: 6.67, NNZs: 4, Bias: 23.244833, T: 3872, Avg. loss: 6.109667
Total training time: 0.00 seconds.
-- Epoch 12
Norm: 6.87, NNZs: 4, Bias: 23.427557, T: 4224, Avg. loss: 6.084979
Total training time: 0.00 

Convergence after 10 epochs took 0.03 seconds
-- Epoch 1
Norm: 5.40, NNZs: 6, Bias: 22.878269, T: 353, Avg. loss: 24.673287
Total training time: 0.00 seconds.
-- Epoch 2
Norm: 5.47, NNZs: 6, Bias: 23.254716, T: 706, Avg. loss: 6.104862
Total training time: 0.00 seconds.
-- Epoch 3
Norm: 5.64, NNZs: 6, Bias: 23.750453, T: 1059, Avg. loss: 5.997466
Total training time: 0.00 seconds.
-- Epoch 4
Norm: 5.46, NNZs: 6, Bias: 23.521868, T: 1412, Avg. loss: 5.903909
Total training time: 0.00 seconds.
-- Epoch 5
Norm: 5.67, NNZs: 6, Bias: 23.598904, T: 1765, Avg. loss: 6.032628
Total training time: 0.00 seconds.
-- Epoch 6
Norm: 5.57, NNZs: 6, Bias: 23.557560, T: 2118, Avg. loss: 6.039860
Total training time: 0.00 seconds.
-- Epoch 7
Norm: 5.53, NNZs: 6, Bias: 23.325954, T: 2471, Avg. loss: 6.021644
Total training time: 0.00 seconds.
-- Epoch 8
Norm: 5.47, NNZs: 6, Bias: 23.815269, T: 2824, Avg. loss: 5.839150
Total training time: 0.00 seconds.
-- Epoch 9
Norm: 5.50, NNZs: 6, Bias: 23.287635, T:

### Análise dos parâmetros

In [ ]:
# Best_estimator: mostra todos os parâmetros com melhor performance apresentado no treinamento
print(grid.best_estimator_)

SGDRegressor(alpha=0.005, average=False, early_stopping=False, epsilon=0.1,
             eta0=0.1, fit_intercept=True, l1_ratio=0.15,
             learning_rate='invscaling', loss='squared_loss', max_iter=1000,
             n_iter_no_change=5, penalty='L1', power_t=0.25, random_state=590,
             shuffle=True, tol=0.001, validation_fraction=0.1, verbose=True,
             warm_start=False)


In [ ]:
# best_params_: mostra os dados da configuração do param_grid, as melhores combinações de performance
print(grid.best_params_)

{'alpha': 0.005, 'eta0': 0.1, 'penalty': 'L1'}


In [ ]:
# cv_results_: mostra toda a variação de testes realizados dado a configuração do param_grid
print(grid.cv_results_)

{'mean_fit_time': array([0.01491537, 0.01922126, 0.00493844, 0.00461211, 0.00320768,
       0.00353839, 0.01417735, 0.01388283, 0.00502379, 0.0057009 ,
       0.00280035, 0.00240018, 0.01532335, 0.01510158, 0.00500028,
       0.00436294, 0.00454144, 0.00360043, 0.01347938, 0.01378477,
       0.00503478, 0.00280015, 0.00431585, 0.0049371 ]), 'std_fit_time': array([0.0107157 , 0.01485713, 0.00444568, 0.00264851, 0.00059872,
       0.00155741, 0.00300514, 0.00214978, 0.00132671, 0.00161544,
       0.00059911, 0.00049001, 0.00978211, 0.00747635, 0.00256969,
       0.00163186, 0.00151872, 0.00080038, 0.0022166 , 0.00117757,
       0.0024659 , 0.00060004, 0.00495488, 0.00770085]), 'mean_score_time': array([0.00050001, 0.00049961, 0.00039973, 0.0006    , 0.00020003,
       0.00029986, 0.00019987, 0.00030005, 0.00049996, 0.00049968,
       0.00059991, 0.00029998, 0.00069997, 0.00049939, 0.00020003,
       0.00049992, 0.00040009, 0.0006999 , 0.00059998, 0.00050008,
       0.00060015, 0.00030005

In [ ]:
# best_score_: mostra o valor da melhor performance dentre todas as combinações, mas representando a combinação vista best_params
# Neste caso para R2
print((grid.best_score_)*100)

65.2386658995281


### <font color = '#8A2BE2'>VAMOS AO QUE HÁ DE MAIS INTERESSANTE NA EXTRAÇÃO DE INFORMAÇÕES DOS ATRIBUTOS</font>

In [ ]:
# Extrair o valor do intercept: momento em que x = 0 e cruza o eixo y
print(grid.best_estimator_.intercept_)

[23.70411494]


In [ ]:
# Extrair os coeficientes de cada variável
pd.DataFrame(X.columns, grid.best_estimator_.coef_)

0
-0.459965     cylinders
 0.507480  displacement
 0.000000    horsepower
-5.764032        weight
 0.022467  accelaration
 2.808219    model_year

### Interpretação dos coeficientes
a) quanto mais cilindros, menor a eficiencia do veículo de mpg(milhas por galão)

b) quanto maior o deslocamento do motor, mais o veículo rende por galão

c) horsepower foi descartado pelo algoritmo. Talvez tenha gerado multicolinearidade

d) quanto mais pesado o veículo menor o rendimento por galão

e) quanto maior a aceleração, maior o rendimento por galão

f) quanto mais novo o modelo do veículo maior o rendimento por galão.

#### É estranho a relação proporcional entre displacement e mpg, visto que são grandezas inversas.
#### Não irei resolver esse problema aqui, por não ser o objetivo do post, mas fica pra você analisar com uma gráfico de correlação.
#### Também é possivel eliminar valores outiliers, o que melhoria os valores dos atributos

## <font color = '#228B22'>CONCLUSÕES</font>

### <font color = '#6495ED'>MELHOR RESULTADO PARA SGDREGRESSOR</font>
-- Epoch 24

Norm (norma): 8.78, 

NNZs (nº de dimensões diferente de zero): 5, 

Bias(viés da variância): 23.704115, 

T (peso aplicado em cada rodada): 9408, 

Avg. loss (custo médio): 6.130518


Convergence after 24 epochs took 0.00 seconds


### MELHORES PARÂMETROS SELECIONADOS PELO GRIDSEARHCV
{'alpha': 0.005, 'eta0': 0.1, 'penalty': 'L1'}

### MELHOR SCORE PARA R2 (%)
65.2386658995281

#### Não era o foco fazer o treinamento e previsão neste notebook, mas sim transmitir o poder do GridSearchCV e sua facilidade de aplicação como suporte a outros avaliadores, como o SGDRegressor.

#### O foco era nos atributos, pois são eles que nos dão os insights para continuar melhorando o treinamento.

# <font color = '#0000FF'>FIM</font>